### PR TITLE
Bal 1427 alert definition v2

### PR DESCRIPTION
**Pull request by liorzam**
## **Type**
enhancement

___

## **Description**
- Implemented dynamic rule generation for alert definitions in `generate-alerts.ts`, including the addition of new imports and functions.
- Defined `AggregateType` constants in `consts.ts` for use across data analytics functionalities.
- Refactored `evaluate-transactions-against-dynamic-rules.ts` to utilize the new `AggregateType` constants and support dynamic aggregation methods.
- Introduced new types in `evaluate-types.ts` to facilitate the dynamic evaluation of transactions.
- Updated Docker Compose scripts in `package.json` for better configuration management.
- Updated the data migrations submodule reference to the latest commit.

___

## **Changes walkthrough**
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>generate-alerts.ts</strong><dd><code>Implement Dynamic Alert Rule Generation</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

services/workflows-service/scripts/alerts/generate-alerts.ts
<li>Added imports for <code>TransactionsAgainstDynamicRulesType</code>, <code>rules</code>, <br><code>PaymentMethod</code>, and <code>AggregateType</code>.<br> <li> Implemented <code>generateRule</code> and <code>getRuleDefinitions</code> functions for dynamic <br>rule generation.<br> <li> Updated <code>generateFakeAlertDefinition</code> to use dynamic rule definitions.

</details>
    

  </td>
  <td><a href="https://github.com/ballerine-io/ballerine/pull/2103/files#diff-65f82c5656fd8455fd436980f8c7d07ace7862374973f5493734b25cb69cce9f">+92/-5</a>&nbsp; &nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>consts.ts</strong><dd><code>Define AggregateType Constants</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

services/workflows-service/src/data-analytics/consts.ts
<li>Defined <code>AggregateType</code> constants for SUM, AVG, COUNT, MAX, and MIN.

</details>
    

  </td>
  <td><a href="https://github.com/ballerine-io/ballerine/pull/2103/files#diff-8a6e7237eb4e05a63b3f704be79ebe2df8571e0d954910a01148821cf5400540">+7/-0</a>&nbsp; &nbsp; &nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>evaluate-transactions-against-dynamic-rules.ts</strong><dd><code>Refactor Transaction Evaluation for Dynamic Aggregation</code>&nbsp; &nbsp; </dd></summary>
<hr>

services/workflows-service/src/data-analytics/data-analytics.repository/evaluate-transactions-against-dynamic-rules.ts
<li>Refactored to use <code>AggregateType</code> from constants.<br> <li> Modified <code>evaluateTransactionsAgainstDynamicRules</code> to support dynamic <br>aggregation.<br> <li> Adjusted query construction for dynamic aggregation.

</details>
    

  </td>
  <td><a href="https://github.com/ballerine-io/ballerine/pull/2103/files#diff-f859a4f61bb031f02c9dff3d48dfd84d62cc540ba416d8007c566fa4c7059ec7">+27/-29</a>&nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>evaluate-types.ts</strong><dd><code>Define Types for Transaction Evaluation</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

services/workflows-service/src/data-analytics/evaluate-types.ts
<li>Introduced types <code>InlineRule</code>, <code>TAggregations</code>, and <br><code>TransactionsAgainstDynamicRulesType</code>.

</details>
    

  </td>
  <td><a href="https://github.com/ballerine-io/ballerine/pull/2103/files#diff-8a86e588fc0c5e5b12d398c83dead0be4d5c4a3a2b5c27212f2b5f5023d10613">+21/-0</a>&nbsp; &nbsp; </td>
</tr>                    
</table></td></tr><tr><td><strong>Configuration changes</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>package.json</strong><dd><code>Update Docker Compose Scripts in package.json</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

services/workflows-service/package.json
<li>Modified <code>compose:up</code> and <code>compose:down</code> scripts to specify <br><code>docker-compose.db.yml</code>.

</details>
    

  </td>
  <td><a href="https://github.com/ballerine-io/ballerine/pull/2103/files#diff-65672d584fae78ff93600b1898b6b95bb50b5e51a25297cd36a840c37cd34b65">+2/-2</a>&nbsp; &nbsp; &nbsp; </td>
</tr>                    
</table></td></tr><tr><td><strong>Dependencies</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>data-migrations</strong><dd><code>Update Data Migrations Submodule Reference</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

services/workflows-service/prisma/data-migrations
- Updated submodule reference for data migrations.

</details>
    

  </td>
  <td><a href="https://github.com/ballerine-io/ballerine/pull/2103/files#diff-9c33eef7eba3ba7ca6006881973adb35d0cbc4048bb7256f945ab6da6f013edf">+1/-1</a>&nbsp; &nbsp; &nbsp; </td>
</tr>                    
</table></td></tr></tr></tbody></table><hr>

<details> <summary><strong>✨ Usage guide:</strong></summary><hr> 

**Overview:**
The `describe` tool scans the PR code changes, and generates a description for the PR - title, type, summary, walkthrough and labels. The tool can be triggered [automatically](https://github.com/Codium-ai/pr-agent/blob/main/Usage.md#github-app-automatic-tools) every time a new PR is opened, or can be invoked manually by commenting on a PR.

When commenting, to edit [configurations](https://github.com/Codium-ai/pr-agent/blob/main/pr_agent/settings/configuration.toml#L46) related to the describe tool (`pr_description` section), use the following template:
```
/describe --pr_description.some_config1=... --pr_description.some_config2=...
```
With a [configuration file](https://github.com/Codium-ai/pr-agent/blob/main/Usage.md#working-with-github-app), use the following template:
```
[pr_description]
some_config1=...
some_config2=...
```

<table><tr><td><details> <summary><strong> Enabling\disabling automation </strong></summary><hr>

- When you first install the app, the [default mode](https://github.com/Codium-ai/pr-agent/blob/main/Usage.md#github-app-automatic-tools) for the describe tool is:
```
pr_commands = ["/describe --pr_description.add_original_user_description=true" 
                         "--pr_description.keep_original_user_title=true", ...]
```
meaning the `describe` tool will run automatically on every PR, will keep the original title, and will add the original user description above the generated description. 

- Markers are an alternative way to control the generated description, to give maximal control to the user. If you set:
```
pr_commands = ["/describe --pr_description.use_description_markers=true", ...]
```
the tool will replace every marker of the form `pr_agent:marker_name` in the PR description with the relevant content, where `marker_name` is one of the following:
  - `type`: the PR type.
  - `summary`: the PR summary.
  - `walkthrough`: the PR walkthrough.

Note that when markers are enabled, if the original PR description does not contain any markers, the tool will not alter the description at all.
        

</details></td></tr>

<tr><td><details> <summary><strong> Custom labels </strong></summary><hr>

The default labels of the `describe` tool are quite generic: [`Bug fix`, `Tests`, `Enhancement`, `Documentation`, `Other`].

If you specify [custom labels](https://github.com/Codium-ai/pr-agent/blob/main/docs/DESCRIBE.md#handle-custom-labels-from-the-repos-labels-page-gem) in the repo's labels page or via configuration file, you can get tailored labels for your use cases.
Examples for custom labels:
- `Main topic:performance` - pr_agent:The main topic of this PR is performance
- `New endpoint` - pr_agent:A new endpoint was added in this PR
- `SQL query` - pr_agent:A new SQL query was added in this PR
- `Dockerfile changes` - pr_agent:The PR contains changes in the Dockerfile
- ...

The list above is eclectic, and aims to give an idea of different possibilities. Define custom labels that are relevant for your repo and use cases.
Note that Labels are not mutually exclusive, so you can add multiple label categories.
Make sure to provide proper title, and a detailed and well-phrased description for each label, so the tool will know when to suggest it.        

</details></td></tr>

<tr><td><details> <summary><strong> Inline File Walkthrough 💎</strong></summary><hr>

For enhanced user experience, the `describe` tool can add file summaries directly to the "Files changed" tab in the PR page.
This will enable you to quickly understand the changes in each file, while reviewing the code changes (diffs).

To enable inline file summary, set `pr_description.inline_file_summary` in the configuration file, possible values are:
- `'table'`: File changes walkthrough table will be displayed on the top of the "Files changed" tab, in addition to the "Conversation" tab.
- `true`: A collapsable file comment with changes title and a changes summary for each file in the PR.
- `false` (default): File changes walkthrough will be added only to the "Conversation" tab.
<tr><td><details> <summary><strong> Utilizing extra instructions</strong></summary><hr>

The `describe` tool can be configured with extra instructions, to guide the model to a feedback tailored to the needs of your project.

Be specific, clear, and concise in the instructions. With extra instructions, you are the prompter. Notice that the general structure of the description is fixed, and cannot be changed. Extra instructions can change the content or style of each sub-section of the PR description.

Examples for extra instructions:
```
[pr_description] 
extra_instructions="""
- The PR title should be in the format: '<PR type>: <title>'
- The title should be short and concise (up to 10 words)
- ...
"""
```
Use triple quotes to write multi-line instructions. Use bullet points to make the instructions more readable.

</details></td></tr>

<tr><td><details> <summary><strong> More PR-Agent commands</strong></summary><hr> 

> To invoke the PR-Agent, add a comment using one of the following commands:  
> - **/review**: Request a review of your Pull Request.   
> - **/describe**: Update the PR title and description based on the contents of the PR.   
> - **/improve [--extended]**: Suggest code improvements. Extended mode provides a higher quality feedback.   
> - **/ask \<QUESTION\>**: Ask a question about the PR.   
> - **/update_changelog**: Update the changelog based on the PR's contents.   
> - **/add_docs** 💎: Generate docstring for new components introduced in the PR.   
> - **/generate_labels** 💎: Generate labels for the PR based on the PR's contents.   
> - **/analyze** 💎: Automatically analyzes the PR, and presents changes walkthrough for each component.   

>See the [tools guide](https://github.com/Codium-ai/pr-agent/blob/main/docs/TOOLS_GUIDE.md) for more details.
>To list the possible configuration parameters, add a **/config** comment.   
 

</details></td></tr>

</table>

See the [describe usage](https://github.com/Codium-ai/pr-agent/blob/main/docs/DESCRIBE.md) page for a comprehensive guide on using this tool.

</details>

### CODE REVIEW

File: services/workflows-service/src/data-analytics/data-analytics.repository/evaluate-transactions-against-dynamic-rules.ts [hunk 1]
```diff
@@ -1,42 +1,29 @@
...
-  MIN = 'MIN',
-}
+const ruleToEvaluateFunction = {
+  [evaluateTransactionsAgainstDynamicRules.name]: (options: InlineRule) => {
+    evaluateTransactionsAgainstDynamicRules(options);
+  },
+};
 
-export const evaluateTransactionsAgainstDynamicRules = async ({
+export async function evaluateTransactionsAgainstDynamicRules({
```
Comment (alonp99) [inline services/workflows-service/src/data-analytics/data-analytics.repository/evaluate-transactions-against-dynamic-rules.ts]: We have a lint rule against this, why'd you change it?

File: services/workflows-service/src/data-analytics/data-analytics.repository/evaluate-transactions-against-dynamic-rules.ts [hunk 2]
```diff
@@ -101,16 +88,27 @@ export const evaluateTransactionsAgainstDynamicRules = async ({
     groupByClause = Prisma.sql`"businessId", "counterpartyOriginatorId"`;
   }
 
-  // test this to allow more aggregation methods
-  const aggregateFunction = Prisma.sql`${havingAggregate}(tr."transactionBaseAmount")`;
+  let havingClause: string = '';
+  switch (havingAggregate) {
```
Comment (Omri-Levy) [inline services/workflows-service/src/data-analytics/data-analytics.repository/evaluate-transactions-against-dynamic-rules.ts:86]:
```suggestion
  let havingClause: string = '';
  
  switch (havingAggregate) {
```

File: services/workflows-service/src/data-analytics/evaluate-types.ts [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+  timeUnit?: 'minutes' | 'hours' | 'days' | 'weeks' | 'months' | 'years';
+  direction?: TransactionDirection;
+  excludedCounterpartyIds?: string[];
+  paymentMethods?: PaymentMethod[];
+  excludePaymentMethods?: boolean;
+  days?: number;
+  amountThreshold?: number;
+  groupByBusiness?: boolean;
+  groupByCounterparty?: boolean;
+};
```
Comment (Omri-Levy) [inline services/workflows-service/src/data-analytics/evaluate-types.ts:21]: Are all those really optional or is this type supposed to represent a union?

File: services/workflows-service/scripts/alerts/generate-alerts.ts [hunk 4]
```diff
@@ -20,6 +27,121 @@ const tags = [
...
+      excludedCounterpartyIds: ['9999999999999999', '999999******9999'],
+
+      paymentMethods: [PaymentMethod.CreditCard],
+      excludePaymentMethods: true,
+
+      timeAmount: 7,
+      timeUnit: 'days',
+
+      amountThreshold: 1000,
+    } as TransactionsAgainstDynamicRulesType,
```
Comment (Omri-Levy) [inline services/workflows-service/scripts/alerts/generate-alerts.ts:86]: All these instances fit the `satisfies` keyword's use-case. 🙂

File: services/workflows-service/src/data-analytics/data-analytics.repository/evaluate-transactions-against-dynamic-rules.ts (hunk 1 above)
Comment (liorzam) [reply services/workflows-service/src/data-analytics/data-analytics.repository/evaluate-transactions-against-dynamic-rules.ts]: than i have to set the mapping in the bottom of the file

File: services/workflows-service/src/data-analytics/evaluate-types.ts (hunk 3 above)
Comment (liorzam) [reply services/workflows-service/src/data-analytics/evaluate-types.ts:21]: @alonp99

File: services/workflows-service/scripts/alerts/generate-alerts.ts (hunk 4 above)
Comment (liorzam) [reply services/workflows-service/scripts/alerts/generate-alerts.ts:86]: i used it and than i faced with issues from prisma

File: services/workflows-service/src/data-analytics/data-analytics.repository/evaluate-transactions-against-dynamic-rules.ts (hunk 1 above)
Comment (liorzam) [reply services/workflows-service/src/data-analytics/data-analytics.repository/evaluate-transactions-against-dynamic-rules.ts]: will change it in my next pr

**Pull request closed without merging**